Implements --machine flag for `devices` command

kika opened the pull request:
## Description

This PR adds `--machine` flag to `devices` command of the Flutter CLI tool. It outputs information about connected devices in a machine-readable indented JSON.

## Related Issues

https://github.com/flutter/flutter/issues/50449

## Tests

I added the following tests:

  1. Test for the new `Device.toJson()` method which verifies that the resulting `Map<String, Object>` representation actually represents the `Device` object as expected.
  1. Test for the `--machine` flag for the `device` command which verifies that the resulting JSON is parseable and represents the same `Device` object.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

<!-- Links -->
[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

kika committed:
Implements --machine flag for `devices` command; See https://github.com/flutter/flutter/issues/50449

jonahwilliams commented on packages/flutter_tools/lib/src/device.dart:
```diff
@@ -490,6 +491,33 @@ abstract class Device {
+    );
+  }
+
+  Future<Map<String, Object>> toJson() async =>
```
This method needs a doc comment, and a unit test for different device types.

kika replied:
@jonahwilliams Could you please elaborate on "different device types"? What is the possible failure scenario to test against?

jonahwilliams replied:
You need to verify that your JSON mapping is correct. Suppose I changed key values randomly, that should make some test fail

jonahwilliams commented on packages/flutter_tools/lib/src/device.dart:
```diff
@@ -490,6 +491,33 @@ abstract class Device {
   }
 
+  static Future<void> printDevicesAsJson(List<Device> devices) async {
+    globals.printStatus(
```
I would move this logic into the devices command since that is the only place it is used

kika replied:
I was actually confused about why `printDevices()` is a static method of the `Device`, so I decided to put mine alongside just to blend in :-)

jonahwilliams commented on packages/flutter_tools/lib/src/commands/devices.dart:
```diff
@@ -42,6 +50,8 @@ class DevicesCommand extends FlutterCommand {
           globals.printStatus('• $diagnostic', hangingIndent: 2);
         }
       }
+    } else if(boolArg('machine')) {
```
nit: space between `if` and `(`

kika replied:
🤦‍♂️

jonahwilliams commented on packages/flutter_tools/lib/src/commands/devices.dart:
```diff
@@ -12,6 +12,14 @@ import '../globals.dart' as globals;
 import '../runner/flutter_command.dart';
 
 class DevicesCommand extends FlutterCommand {
```
The Devices command also needs a test for this new functionality, that passing the flag outputs parsable JSON. You'll need to mock the device manager with a fake device so you know what to expect

jonahwilliams commented:
Seems like a reasonable change, but you'll need test coverage

kika committed:
Addresses reviewer concerns at https://github.com/flutter/flutter/pull/50581 ; except for tests

kika committed:
Added tests; _MockDevice moved to separate file for sharing between different tests

kika committed:
Renamed _MockDevice into MockDevice in other tests for DRY

kika committed:
Cosmetics

kika commented:
@jonahwilliams rebased and ready for review.

kika committed:
Removed blank line at the end of the file

kika committed:
Fixed changed method signature after rebase

kika committed:
Removed dependency on the main source and replaced with the original Dart package: analyzer complaint

kika committed:
More excessive unnecessary type declarations

jmagman commented:
There's also a test failure:
```
00:12 +596 -1: test/general.shard/forbidden_imports_test.dart: no unauthorized imports of dart:convert [E]                                                                                             
  lib/src/commands/devices.dart imports 'dart:convert'; import 'lib/src/convert.dart' instead
  package:test_api                                                           fail
  org-dartlang-app:///test/general.shard/forbidden_imports_test.dart 154:13  main.<fn>
```

jonahwilliams commented on packages/flutter_tools/test/commands.shard/hermetic/devices_test.dart:
```diff
@@ -36,6 +37,18 @@ void main() {
+      await createTestCommandRunner(command).run(<String>['devices', '--machine']);
+      expect(
+        json.decode(testLogger.statusText),
+        mockDevices.map((MockDeviceJson d) => d.json)
```
It would be better if you could define the values you are asserting on within the test. As is it is difficult to tell what is doing.

kika replied:
@jonahwilliams I use these values in two different tests and that's why I moved them into common `test/src/mock_devices.dart`. Should I just blindly copy-paste here or just providing a comment is enough? For me, it just takes pressing F12 on `MockDeviceJson` to jump to that file and see everything.

jonahwilliams replied:
Ahh I see you have this already, my b

jonahwilliams commented on packages/flutter_tools/test/commands.shard/hermetic/devices_test.dart:
```diff
@@ -66,3 +79,12 @@ class MockProcessManager extends Mock implements ProcessManager {
+  Future<List<Device>> getAllConnectedDevices() {
+    return Future<List<Device>>.value(mockDevices.map((MockDeviceJson d) => d.dev).toList());
+  }
+}
```
```suggestion
}

```

kika replied:
Hm, interesting. I've seen complaints from the analyzer about blank lines at the end.

jonahwilliams commented on packages/flutter_tools/test/src/mock_devices.dart:
```diff
@@ -0,0 +1,96 @@
+
+  final MockDevice dev;
+  final Map<String, Object> json;
+}
```
```suggestion
}

```

jonahwilliams commented on packages/flutter_tools/lib/src/commands/devices.dart:
```diff
@@ -12,6 +13,14 @@ import '../globals.dart' as globals;
 
 class DevicesCommand extends FlutterCommand {
+
+  DevicesCommand() {
```
This command also needs a test, see the commands.shard for some examples

kika replied:
@jonahwilliams I'm confused. I didn't add this command, I just implemented the constructor (the command did have a default constructor before). And I have tests for the only thing I've added: `--machine` flag. 

Also, the default functionality of the `devices` command does have tests in `test/commands.shard/hermetic/devices_test.dart`

jonahwilliams replied:
(Sorry for the delay)

Right, but you need a test that invokes the --machine and then verifies the output as expected. There are some example tests in tests/commands.shard/permeable/**

kika committed:
Add blank line at the end

Co-Authored-By: Jonah Williams <jonahwilliams@google.com>

kika committed:
Add blank line at the end

Co-Authored-By: Jonah Williams <jonahwilliams@google.com>

kika committed:
Addresses the test failure due to unauthorized import

jonahwilliams commented on packages/flutter_tools/test/src/mock_devices.dart:
```diff
@@ -0,0 +1,96 @@
+];
+
+/// Fake device to test `devices` command
+class MockDevice extends Device {
```
Rename this to `FakeDevice` since it doesn't implement the Mock interface anymore

jonahwilliams commented on packages/flutter_tools/test/src/mock_devices.dart:
```diff
@@ -0,0 +1,96 @@
+}
+
+/// Combines fake device with its canonical JSON representation
+class MockDeviceJson {
```
This isn't a mock, could you instead place the expectations inline?

like `expect(x.toJson(), { .... })`

kika replied:
It's the same problem as in https://github.com/flutter/flutter/pull/50581#pullrequestreview-365904285
I use this data twice in two different tests and to put expectations inline I need to copy-paste twice, which may cause issues down the road if someone decides to change that.

kika replied:
For now I'll just rename it to `FakeDeviceJsonData`. lmk what you think. This is indeed not a mock, just data container.

jonahwilliams replied:
Please put the test expectations in the test. It may involve more copying around, but it is much easier to understand that the test expectations are important. Tests aren't about DRY, its about documentation. For example, I wouldn't write a test like `expect(device.toJson(), equals(device.toJson())`

kika replied:
Delays now on me, need to take care of my team more in these WFH days. I will try to finish over the weekend. 
Writing tests for the commands is not a very welcoming experience I'd say :-) A lot of infrastructure needs to be created just to test `device --machine`.

jonahwilliams commented on packages/flutter_tools/test/commands.shard/hermetic/devices_test.dart:
```diff
@@ -66,3 +79,12 @@ class MockProcessManager extends Mock implements ProcessManager {
   }
 }
+
+class _MockDeviceManager extends DeviceManager {
```
nit: renamed to MockDeviceManager

kika replied:
I guess it should also become `FakeDeviceManager`

jonahwilliams replied:
Yes, if its not extends Mock it is a Fake

kika replied:
@jonahwilliams I guess it's ready for review

jonahwilliams commented:
Sorry for the delay on reviewing this - it got a little crazy here the last few days. I left some comments, this only needs some minor changes

kika committed:
Rename Mock* what are not mocks to Fake*

kika committed:
Update packages/flutter_tools/lib/src/device.dart

Co-Authored-By: Jonah Williams <jonahwilliams@google.com>

kika committed:
Merge branch 'kika/machine-flag-devices' of https://github.com/kika/flutter into kika/machine-flag-devices

kika committed:
Added `devices --machine` permeable test

kika committed:
supportsHardwareRendering is being called on web devices which are not local emulators; causes assert failure

kika committed:
Copied expected data into the call into the site of `expect()` function call itself; Renamed MockDeviceManager to _FakeDeviceManager

kika committed:
Merge and resolve conflicts

kika committed:
Fixed invalid override (merge?)

kika committed:
Fixed tests:
DeviceManager interface changed
Rename of the field in the device JSON representation
--
Uh-oh, it's getting harder to keep up with master...

jonahwilliams reviewed (approved):
LGTM
